File: jurt/com/sun/star/lib/uno/protocols/urp/Cache.java
/* -*- Mode: Java; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */

package com.sun.star.lib.uno.protocols.urp;

import java.util.HashMap;

/**
 * An LRU cache for arbitrary objects.
 *
 * <p>This class is not synchronized, as any necessary synchronization will already
 * take place in the client.</p>
 */
final class Cache {
    /**
     * Create a cache.
     *
     * @param size the maximum cache size, must be between 0, inclusive, and
     * NOT_CACHED, exclusive.
     */
    public Cache(int size) {
        maxSize = size;
    }

    public int add(boolean[] found, Object content) {
        Entry e = map.get(content);
        found[0] = e != null;
        if (e == null) {
            if (map.size() < maxSize) {
                // There is still room for a new entry at the front:
                e = new Entry(content, map.size(), null, first);
                if (first == null) {
                    last = e;
                } else {
                    first.prev = e;
                }
                first = e;
            } else if (last != null) {
                // Take last entry out and recycle as new front:
                map.remove(last.content);
                e = last;
                e.content = content;
                if (first != last) {
                    // Reached only if maxSize > 1:
                    last = last.prev;
                    last.next = null;
                    e.prev = null;
                    e.next = first;
                    first.prev = e;
                    first = e;
                }
            } else {
                // Reached if maxSize == 0:
                return NOT_CACHED;
            }
            map.put(content, e);
        } else if (e != first) {
            // Move to front (reached only if maxSize > 1):
            e.prev.next = e.next;
            if (e.next == null) {
                last = e.prev;
            } else {
                e.next.prev = e.prev;
            }
            e.prev = null;
            e.next = first;
            first.prev = e;
            first = e;
        }
        return e.index;
    }

    public static final int NOT_CACHED = 0xFFFF;

    private static final class Entry {
        public Entry(Object content, int index, Entry prev, Entry next) {
            this.content = content;
            this.index = index;
            this.prev = prev;
            this.next = next;
        }

        public Object content;
        public int index;
        public Entry prev;
        public Entry next;
    }

    // first/last form a list of 0 to maxSize entries, most recently used first;
    // map contains the same entries; each entry has a unique index in the range
    // 0 to maxSize - 1
    private final int maxSize;
    private final HashMap<Object, Entry> map = new HashMap<Object, Entry>(); // from Object to Entry
    private Entry first = null;
    private Entry last = null;
}

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */

File: sc/workben/cache.py
# -*- Mode: python; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

import sys
import struct

column_block_type = {
        0 : "empty block",
        1 : "numeric block",
        2 : "string block"
        }


def parse_block(data, index):
    start_row = struct.unpack('Q', data[index:index+8])[0]
    index += 8
    data_size = struct.unpack('Q', data[index:index+8])[0]
    index += 8
    block_type = struct.unpack('B', data[index:index+1])[0]
    index += 1
    vals = {}
    if block_type == 1:
        # numeric block
        for i in range(data_size):
            vals[start_row + i] = struct.unpack('d', data[index:index+8])[0]
            index += 8
    elif block_type == 2:
        # string block
        for i in range(data_size):
            str_length = struct.unpack('i', data[index:index+4])[0]
            index += 4
            vals[start_row + i] = data[index:index+str_length].decode("utf-8")
            index += str_length
    elif block_type == 3:
        # formula block
        read_rows = 0
        while read_rows < data_size:
            formula_group_size = struct.unpack('Q', data[index:index+8])[0]
            index += 8
            str_length = struct.unpack('i', data[index:index+4])[0]
            index += 4
            vals[start_row + read_rows] = (data[index:index+str_length].decode("utf-8"), "formula group length %i"% formula_group_size)
            read_rows += formula_group_size
            index += str_length

    return index, data_size, vals


def parse_column(data, index):
    column_index = struct.unpack('Q', data[index:index+8])[0]
    index += 8
    column_entries = struct.unpack('Q', data[index:index+8])[0]
    index += 8
    imported_columns = 0
    column_values = {}
    while imported_columns < column_entries:
        index, block_size, vals = parse_block(data, index)
        imported_columns += block_size
        column_values.update(vals)
    return index, column_values

def parse_columns(data, index):
    column_count = struct.unpack('Q', data[index:index+8])[0]
    index += 8
    columns = {}
    for i in range(column_count):
        index, column_values = parse_column(data, index)
        columns[i] = column_values

    return columns

def main():
    filename = sys.argv[1]
    with open(filename, "rb") as f:
        content = f.read()
        index = 0
        columns = parse_columns(content, index)
        print(columns)

if __name__ == "__main__":
    main()

# vim:set shiftwidth=4 softtabstop=4 expandtab: */
